python/tgn.py

```diff
@@ -0,0 +1,695 @@
+import enum
+from typing import Dict, Tuple, Set, List, Any
+
+import mgp
+
+import numpy as np
+import torch
+import torch.nn as nn
+from mage.tgn.constants import (
+    TGNLayerType,
+    MessageFunctionType,
+    MessageAggregatorType,
+    MemoryUpdaterType,
+)
+from mage.tgn.definitions.tgn import TGN
+from dataclasses import dataclass
+
+
+from mage.tgn.helper.simple_mlp import MLP
+
+from mage.tgn.definitions.instances import (
+    TGNGraphSumEdgeSelfSupervised,
+    TGNGraphSumSupervised,
+    TGNGraphAttentionSupervised,
+    TGNGraphAttentionEdgeSelfSupervised,
+)
+
+
+###################
+# params and classes
+##################
+
+
+class TGNParameters:
+    NUM_OF_LAYERS = "num_of_layers"
+    LAYER_TYPE = "layer_type"
+    MEMORY_DIMENSION = "memory_dimension"
+    TIME_DIMENSION = "time_dimension"
+    NUM_EDGE_FEATURES = "num_edge_features"
+    NUM_NODE_FEATURES = "num_node_features"
+    MESSAGE_DIMENSION = "message_dimension"
+    NUM_NEIGHBORS = "num_neighbors"
+    EDGE_FUNCTION_TYPE = "edge_message_function_type"
+    MESSAGE_AGGREGATOR_TYPE = "message_aggregator_type"
+    MEMORY_UPDATER_TYPE = "memory_updater_type"
+    NUM_ATTENTION_HEADS = "num_attention_heads"
+    DEVICE = "device"
+
+
+class OptimizerParameters:
+    LEARNING_RATE = "learning_rate"
+    WEIGHT_DECAY = "weight_decay"
+
+
+class LearningType(enum.Enum):
+    Supervised = "supervised"
+    SelfSupervised = "self_supervised"
+
+
+@dataclass
+class QueryModuleTGN:
+    config: Dict[str, Any]
+    tgn: TGN
+    criterion: nn.BCELoss
+    optimizer: torch.optim.Adam
+    device: torch.device
+    m_loss: List[float]
+    mlp: MLP
+
+
+@dataclass
+class QueryModuleTGNBatch:
+    current_batch_size: int
+    sources: np.array
+    destinations: np.array
+    timestamps: np.array
+    edge_idxs: np.array
+    node_features: Dict[int, torch.Tensor]
+    edge_features: Dict[int, torch.Tensor]
+    batch_size: int
+    labels: np.array
+
+
+##############################
+# global tgn training variables
+##############################
+
+# used in negative sampling for self_supervised
+all_edges: Set[Tuple[int, int]] = set()
+
+# to get all embeddings
+all_embeddings: Dict[int, np.array] = {}
+
+query_module_tgn: QueryModuleTGN
+query_module_tgn_batch: QueryModuleTGNBatch
+
+tgn_learning_type = None
+tgn_training_losses = np.empty(0)
```

Comment:
Rather than keeping them here exposed in the module, why not create a singleton in `/tgn` with appropriate methods that are updating this

Reply:
This is updated now, you can check how I implemented it, but I am not sure what are you aiming at.